chia/util/file_keyring.py

```diff
@@ -0,0 +1,438 @@
+import base64
+import fasteners
+import os
+import shutil
+import sys
+import threading
+import yaml
+
+from chia.util.default_root import DEFAULT_KEYS_ROOT_PATH
+from contextlib import contextmanager
+from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305  # pyright: reportMissingModuleSource=false
+from functools import wraps
+from hashlib import pbkdf2_hmac
+from pathlib import Path
+from secrets import token_bytes
+from typing import Optional
+from watchdog.events import FileSystemEventHandler
+from watchdog.observers import Observer
+
+
+SALT_BYTES = 16  # PBKDF2 param
+NONCE_BYTES = 12  # ChaCha20Poly1305 nonce is 12-bytes
+HASH_ITERS = 100000  # PBKDF2 param
+CHECKBYTES_VALUE = b"5f365b8292ee505b"  # Randomly generated
+MAX_SUPPORTED_VERSION = 1  # Max supported file format version
+
+
+class FileKeyringLockTimeout(Exception):
+    pass
+
+
+def loads_keyring(method):
+    """
+    Decorator which lazily loads the FileKeyring data
+    """
+
+    @wraps(method)
+    def inner(self, *args, **kwargs):
+        # Watchdog's event dispatch timing is unreliable on macOS. Force a file modification time check for macOS.
+        if sys.platform == "darwin":
+            self.check_if_keyring_file_modified()
+
+        # Check the outer payload for 'data', and check if we have a decrypted cache (payload_cache)
+        with self.load_keyring_lock:
+            if (self.has_content() and not self.payload_cache) or self.needs_load_keyring:
+                self.load_keyring()
+        return method(self, *args, **kwargs)
+
+    return inner
+
+
+@contextmanager
+def acquire_writer_lock(lock_path: Path, timeout=5, max_iters=6):
+    lock = fasteners.InterProcessReaderWriterLock(str(lock_path))
+    result = None
+    for i in range(0, max_iters):
+        if lock.acquire_write_lock(timeout=timeout):
+            yield  # <----
```

Comment:
@paninaro what is the yield used for here ? i thought this was usually used to create generators in python